compiler/rustc_interface/src/util.rs

```diff
@@ -166,19 +166,21 @@ pub fn run_in_thread_pool_with_globals<F: FnOnce() -> R + Send, R: Send>(
 unsafe fn handle_deadlock() {
     let registry = rustc_rayon_core::Registry::current();
 
-    let context = tls::get_tlv();
-    assert!(context != 0);
-    rustc_data_structures::sync::assert_sync::<tls::ImplicitCtxt<'_, '_>>();
-    let icx: &tls::ImplicitCtxt<'_, '_> = &*(context as *const tls::ImplicitCtxt<'_, '_>);
+    // We do not need to copy the query ImplicitCtxt since this deadlock handler is not part of a
+    // normal query invocation.
 
     let session_globals = rustc_span::with_session_globals(|sg| sg as *const _);
     let session_globals = &*session_globals;
+
+    // Extend the lifetime of the GlobalCtxt so the new thread can know of it.
+    // The current thread will not free it, it is deadlocked.
+    let tcx: &'static GlobalCtxt<'static> =
+        tls::with(|tcx| &*(*tcx as *const GlobalCtxt<'_>).cast());
```

Comment:
This is incorrect since it's only set on one of Rayon's threads. The deadlock handler can be called on any thread. The current state of things is also broken to be fair.

We can not use the Rayon "Task Local Value" either, since the deadlock handler can also be called from a thread with no rustc task is running. My solution to this was to point all threads to a lock which holds the `GlobalCtxt` when available.  It was reverted in https://github.com/rust-lang/rust/pull/74969.